crates/provider/src/new.rs

```diff
@@ -91,13 +121,171 @@ impl<N, T> RootProviderInner<N, T> {
 #[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
 #[auto_impl::auto_impl(&, &mut, Rc, Arc, Box)]
 pub trait Provider<N: Network, T: Transport + Clone = BoxTransport>: Send + Sync {
+    /// Returns the root provider.
+    fn root(&self) -> &RootProvider<N, T>;
```

Comment:
what's the downside of this? when would you use a custom provider as your root provider?

if we go with this, we should update the `ProviderBuilder` to not accept anything other than `RootProvider` as the final provider (i.e. prob just remove `ProviderBuilder::provider`)